src/software/embedded/gpio_char_dev.cpp

```diff
@@ -0,0 +1,111 @@
+#include "software/embedded/gpio_char_dev.h"
+
+#include <linux/gpio.h>
+
+#include "software/logger/logger.h"
+
+GpioCharDev::GpioCharDev(int gpio_number, GpioDirection direction,
+                         GpioState initial_state, std::string char_dev_path)
+{
+    int fd = open(char_dev_path.c_str(), O_RDONLY);
+    if (fd < 0)
+    {
+        LOG(FATAL) << "Could not open GPIO chip device";
+    }
+
+    struct gpiohandle_request req;
+    req.lineoffsets[0] = gpio_number;
+    req.lines          = 1;
+
+    switch (direction)
+    {
+        case GpioDirection::OUTPUT:
+        {
+            req.flags = GPIOHANDLE_REQUEST_OUTPUT;
+            break;
+        }
+        case GpioDirection::INPUT:
+        {
+            req.flags = GPIOHANDLE_REQUEST_INPUT;
+            break;
+        }
+        default:
+        {
+            LOG(FATAL) << "Invalid GPIO direction: " << direction;
+        }
+    }
+
+    req.default_values[0] = parseGpioState(initial_state);
+
+    int ret = ioctl(fd, GPIO_GET_LINEHANDLE_IOCTL, &req);
+    if (ret < 0)
+    {
+        LOG(FATAL) << "Could not get GPIO line handle for GPIO " << gpio_number;
+    }
+    close(fd);
+
+    gpio_fd = req.fd;
+
+    LOG(DEBUG) << "GPIO " << gpio_number << " online";
+}
+
+void GpioCharDev::setValue(GpioState state)
+{
+    struct gpiohandle_data data;
+    data.values[0] = parseGpioState(state);
+
+    int ret = ioctl(gpio_fd, GPIOHANDLE_SET_LINE_VALUES_IOCTL, &data);
+    if (ret < 0)
+    {
+        LOG(FATAL) << "Could not set GPIO value";
+    }
+}
+
+GpioCharDev::~GpioCharDev()
+{
+    close(gpio_fd);
+}
+
+GpioState GpioCharDev::getValue()
+{
+    struct gpiohandle_data data;
+    int ret = ioctl(gpio_fd, GPIOHANDLE_GET_LINE_VALUES_IOCTL, &data);
+
+    if (ret < 0)
+    {
+        LOG(FATAL) << "Could not get GPIO value";
+    }
+
+    switch (data.values[0])
+    {
+        case 0:
+        {
+            return GpioState::LOW;
+        }
+        case 1:
+        {
+            return GpioState::HIGH;
+        }
+    }
+
+    LOG(FATAL) << "Unable to parse GPIO value";
+    return GpioState::LOW;
```

Comment:
I am assuming you are trying to set some default value prior to shutting down the process with the LOG(FATAL). However, could terminating the process prior to setting the state to low leave the GPIO in an undetermined state?

Reply:
The intention is to crash here. I need this return to make the compiler happy. Other parts of our code (if you're curious, look at `std::signal` catches in `thunderloop.cpp`) that catch these `FATAL` calls and safely shuts down

Reply:
Could use `__builtin_unreachable()`

Reply:
Wait, maybe not a good idea to use that. https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html


`If control flow reaches the point of the __builtin_unreachable, the program is undefined. `

Reply:
Ok